In [1]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [99]:
cars_data = pd.read_csv("Toyota_corolla_modified.csv", na_values = ['??','????'])
print(cars_data)
print(cars_data.info())

      Price   Age       KM FuelType    HP  MetColor  Automatic    CC  Doors  \
0      6500  23.0  46986.0      NaN  90.0       1.0          0  2000      3   
1     13750  23.0      NaN   Diesel  90.0       NaN          0  2000  three   
2     13950  24.0  41711.0   Diesel   NaN       1.0          0  2000      3   
3     14950   NaN  48000.0   Diesel  90.0       0.0          0  2000      3   
4     13750  30.0  38500.0   Diesel  90.0       0.0          0  2000      3   
...     ...   ...      ...      ...   ...       ...        ...   ...    ...   
1431   7500  69.0  20544.0   Petrol  86.0       1.0          0  1300      3   
1432  10845  72.0  19000.0   Petrol  86.0       0.0          0  1300      3   
1433   8500  71.0  17016.0   Petrol  86.0       0.0          0  1300      3   
1434   7250  70.0  16916.0   Petrol  86.0       1.0          0  1300      3   
1435   6950  76.0      NaN   Petrol   NaN       0.0          0  1600      5   

      Weight  
0       1165  
1       1165  
2     

In [100]:
cars_data2 = cars_data.copy()
cars_data3 = cars_data2.copy()

# IDENTIFYING MISSING VALUES

### Dataframe.isna().sum()     OR Dataframe.isnull().sum()
gives total count of missing values in each column

In [101]:
print(cars_data2.isna().sum())
print("\n")
print(cars_data2.isnull().sum())

Price         0
Age           8
KM            5
FuelType      5
HP           13
MetColor      5
Automatic     0
CC            0
Doors         0
Weight        0
dtype: int64


Price         0
Age           8
KM            5
FuelType      5
HP           13
MetColor      5
Automatic     0
CC            0
Doors         0
Weight        0
dtype: int64


we need to check whether a particular row has only 1 or multiple NaN values

If a single row has multiple NaN values, it is better to discard the entire row

if a single row has only 1 NaN value, It is better to fill that NaN value logically

SO, Subsetting rows that have one or more missing values
## CHECKING THE MISSING DATA

### variable = Dataframe[Dataframe.isnull().any(axis = 1)]
Thif function represents return all rows whereatleast 1 value is missing

axis = 0 (for row) or 1 (for column)

In [102]:
missing = cars_data2[cars_data2.isnull().any(axis =1)]
print(missing)
print("\n")
print(missing.info())
print("\n")
print(missing.shape)

      Price   Age        KM FuelType     HP  MetColor  Automatic    CC  Doors  \
0      6500  23.0   46986.0      NaN   90.0       1.0          0  2000      3   
1     13750  23.0       NaN   Diesel   90.0       NaN          0  2000  three   
2     13950  24.0   41711.0   Diesel    NaN       1.0          0  2000      3   
3     14950   NaN   48000.0   Diesel   90.0       0.0          0  2000      3   
9     12950  23.0       NaN   Diesel   69.0       0.0          0  1900      3   
12    19600   NaN   32189.0   Petrol  192.0       0.0          0  1800      3   
19    16950  30.0   64359.0   Petrol  110.0       NaN          0  1600      3   
23    16950  28.0   32220.0   Petrol    NaN       1.0          0  1600      3   
28    16950  28.0   44142.0   Petrol  110.0       NaN          0  1600      3   
44    16950  22.0  100250.0      NaN   90.0       0.0          0  2000      5   
47    15800   NaN   75048.0   Petrol   97.0       1.0          0  1400      5   
169   17795   9.0       NaN 

### two approaches to fill the missing values
1. ( for numerical variable ) fill missing values by mean/median
2. ( for categorical variable ) fill missing values with class with maximum count (modal value... the category with highest frequency)

look at description to know whether numerical variables should be inputed with mean or median
### DataFrame.describe().... (For numerical variables)
generates descriptive statistics that summarize the central tendency, dispersion and shape of dataset's distribution excluding NaN values

it gives:
1. count (No. of observations considered)
2. Mean
3. Std deviation
4. 5 Number summary (Min, 25%, 50%, 75%, Max)
Note : it does not give the value of the top and bottom whisker of the box plot
### DataFrame.describe(include = "O").... (For categorical variables)... (O stands for Object)
it gives:
1. count (No. of observations considered)
2. unique (No. of unique elements .... (integer))
3. top (category with most frequency)
4. frequency (corresponding frequency of the top)

In [103]:
cars_data2.describe()

Price          Age             KM           HP     MetColor  \
count   1436.000000  1428.000000    1431.000000  1423.000000  1431.000000   
mean   10725.949861    56.027311   68672.035639   101.502460     0.674354   
std     3627.944090    18.585810   37440.107144    15.001494     0.468780   
min     4350.000000     1.000000       1.000000    69.000000     0.000000   
25%     8450.000000    44.000000   43000.000000    88.000000     0.000000   
50%     9900.000000    61.000000   63451.000000   110.000000     1.000000   
75%    11950.000000    70.000000   87178.500000   110.000000     1.000000   
max    32500.000000    80.000000  243000.000000   192.000000     1.000000   

         Automatic           CC      Weight  
count  1436.000000  1436.000000  1436.00000  
mean      0.056407  1566.827994  1072.45961  
std       0.230786   187.182436    52.64112  
min       0.000000  1300.000000  1000.00000  
25%       0.000000  1400.000000  1040.00000  
50%       0.000000  1600.000000  1070.00000  
75%       0.000000  1600.000000  1085.00000  
max       1.000000  2000.000000  1615.00000

#### Although "MetColor" and "Automatic" are included in numerical category ... they belong to categorical variables 
#### 0 = False,..... 1 = True

In [104]:
cars_data2.describe(include = 'O')

FuelType Doors
count      1431  1436
unique        3     7
top      Petrol     5
freq       1262   672

#### From numerical variables, missing values are present in Age, KM and HP

#### Age
mean: 56 months

median: 61 months

As there is no significant deviation from the Mean, we can replace the missing values with mean

#### KM
mean: 68672.035639

median: 63451.000000

As there is a significant difference of approx. 5000 KMs between mean and median, it is better to replace missing values with median

#### HP
mean: 101.502460	

median: 110.00000

As there is no dignificant difference  between the mean and median, we can replace the missing values with mean

# FILLING THE MISSING VALUES OF NUMERICAL VARIABLES

### Imputing missing values of 'Age' 
#### dataseries.fillna(value to be filled with, inplace= True/False)
to fill NaN values with a specified value

In [105]:
cars_data2["Age"].mean()

56.02731092436975

In [106]:
cars_data2["Age"].fillna(cars_data2["Age"].mean(),
                        inplace = True)

In [107]:
missing = cars_data2[cars_data2.isnull().any(axis =1)]
print(missing)
print(missing.shape)

      Price   Age        KM FuelType     HP  MetColor  Automatic    CC  Doors  \
0      6500  23.0   46986.0      NaN   90.0       1.0          0  2000      3   
1     13750  23.0       NaN   Diesel   90.0       NaN          0  2000  three   
2     13950  24.0   41711.0   Diesel    NaN       1.0          0  2000      3   
9     12950  23.0       NaN   Diesel   69.0       0.0          0  1900      3   
19    16950  30.0   64359.0   Petrol  110.0       NaN          0  1600      3   
23    16950  28.0   32220.0   Petrol    NaN       1.0          0  1600      3   
28    16950  28.0   44142.0   Petrol  110.0       NaN          0  1600      3   
44    16950  22.0  100250.0      NaN   90.0       0.0          0  2000      5   
169   17795   9.0       NaN   Petrol   98.0       1.0          0  1400      4   
170   18245   9.0       NaN   Petrol  110.0       1.0          0  1600      5   
217   10950  41.0   75697.0   Petrol    NaN       1.0          0  1400      3   
263   12750  40.0   50640.0 

### Imputing missing values of 'KM'

In [108]:
cars_data2["KM"].median()

63451.0

In [109]:
cars_data2["KM"].fillna(cars_data2["KM"].median(),
                        inplace = True)

In [110]:
missing = cars_data2[cars_data2.isnull().any(axis =1)]
print(missing)
print(missing.shape)

      Price   Age        KM FuelType     HP  MetColor  Automatic    CC  Doors  \
0      6500  23.0   46986.0      NaN   90.0       1.0          0  2000      3   
1     13750  23.0   63451.0   Diesel   90.0       NaN          0  2000  three   
2     13950  24.0   41711.0   Diesel    NaN       1.0          0  2000      3   
19    16950  30.0   64359.0   Petrol  110.0       NaN          0  1600      3   
23    16950  28.0   32220.0   Petrol    NaN       1.0          0  1600      3   
28    16950  28.0   44142.0   Petrol  110.0       NaN          0  1600      3   
44    16950  22.0  100250.0      NaN   90.0       0.0          0  2000      5   
217   10950  41.0   75697.0   Petrol    NaN       1.0          0  1400      3   
263   12750  40.0   50640.0   Petrol    NaN       1.0          0  1600      5   
294   13950  42.0   42560.0   Petrol    NaN       1.0          0  1600      5   
393    4450  56.0  129155.0   Diesel    NaN       0.0          0  1800      5   
537    9550  54.0   46856.0 

### Imputing missing values of 'HP'

In [111]:
cars_data2["HP"].mean()

101.5024595924104

In [112]:
cars_data2["HP"].fillna(cars_data2["HP"].mean(),
                        inplace = True)

In [113]:
missing = cars_data2[cars_data2.isnull().any(axis =1)]
print(missing)
print(missing.shape)

      Price   Age        KM FuelType     HP  MetColor  Automatic    CC  Doors  \
0      6500  23.0   46986.0      NaN   90.0       1.0          0  2000      3   
1     13750  23.0   63451.0   Diesel   90.0       NaN          0  2000  three   
19    16950  30.0   64359.0   Petrol  110.0       NaN          0  1600      3   
28    16950  28.0   44142.0   Petrol  110.0       NaN          0  1600      3   
44    16950  22.0  100250.0      NaN   90.0       0.0          0  2000      5   
641    8950  68.0  119500.0   Petrol  110.0       NaN          0  1600      5   
648    8000  64.0  116000.0      NaN   86.0       0.0          0  1300      5   
718    9950  63.0   89000.0   Petrol  110.0       NaN          0  1600      5   
1122   7460  80.0  111392.0      NaN  110.0       0.0          0  1598      3   
1376   7800  73.0   56566.0      NaN  110.0       1.0          0  1600      3   

      Weight  
0       1165  
1       1165  
19      1105  
28      1120  
44      1255  
641     1070  
648

### CHECKING THE MISSING DATA NOW 

In [114]:
print(cars_data2.isnull().sum())

Price        0
Age          0
KM           0
FuelType     5
HP           0
MetColor     5
Automatic    0
CC           0
Doors        0
Weight       0
dtype: int64


# FILLING THE MISSING VALUES OF CATEGORICAL VARIABLES

In [18]:
print(cars_data2["FuelType"].value_counts())
print("\n")
print(cars_data2["MetColor"].value_counts())

Petrol    1262
Diesel     153
CNG         16
Name: FuelType, dtype: int64


1.0    965
0.0    466
Name: MetColor, dtype: int64


Modal value of FuelType is : Petrol

Modal value of MetColor is : 1.0

#### modal value can also be found out by below method:
Since it value_counts() returns a series of unique value in decending order of the frequency... the first element is always the modal value

In [58]:
print(cars_data2["FuelType"].value_counts().index[0])
print(cars_data2["MetColor"].value_counts().index[0])

Petrol
1.0


### Imputing missing values of 'FuelType' 

In [116]:
cars_data2["FuelType"].fillna(cars_data2["FuelType"].value_counts().index[0],
                        inplace = True)

## Here we are not specifying 'petrol' directly... so that the code is generallized... need not be tweaked later if reqd

In [117]:
missing = cars_data2[cars_data2.isnull().any(axis =1)]
print(missing)

     Price   Age        KM FuelType     HP  MetColor  Automatic    CC  Doors  \
1    13750  23.0   63451.0   Diesel   90.0       NaN          0  2000  three   
19   16950  30.0   64359.0   Petrol  110.0       NaN          0  1600      3   
28   16950  28.0   44142.0   Petrol  110.0       NaN          0  1600      3   
641   8950  68.0  119500.0   Petrol  110.0       NaN          0  1600      5   
718   9950  63.0   89000.0   Petrol  110.0       NaN          0  1600      5   

     Weight  
1      1165  
19     1105  
28     1120  
641    1070  
718    1085  


### Imputing missing values of 'MetColor'

In [25]:
cars_data2["MetColor"].mode()[0]

1.0

In [119]:
#cars_data2["MetColor"].fillna(cars_data2["MetColor"].mode().index[0],inplace = True)
cars_data2["MetColor"].fillna(cars_data2["MetColor"].mode().index[0], inplace = True)
## when using mode(), we need to index it as it returns output with index 0... i.e. value of 1st index

In [120]:
missing = cars_data2[cars_data2.isnull().any(axis =1)]
print(missing)

Empty DataFrame
Columns: [Price, Age, KM, FuelType, HP, MetColor, Automatic, CC, Doors, Weight]
Index: []


### CHECKING THE MISSING DATA NOW 

In [121]:
print(cars_data2.isnull().sum())

Price        0
Age          0
KM           0
FuelType     0
HP           0
MetColor     0
Automatic    0
CC           0
Doors        0
Weight       0
dtype: int64


# FILLING BOTH NUMERICAL AND CATEGORICAL VARIABLES AT A SINGLE GO

# USING LAMBDA FUNCTION

In [182]:
print(cars_data3)
print(cars_data3.info())

      Price   Age       KM FuelType    HP  MetColor  Automatic    CC  Doors  \
0      6500  23.0  46986.0      NaN  90.0       1.0          0  2000      3   
1     13750  23.0      NaN   Diesel  90.0       NaN          0  2000  three   
2     13950  24.0  41711.0   Diesel   NaN       1.0          0  2000      3   
3     14950   NaN  48000.0   Diesel  90.0       0.0          0  2000      3   
4     13750  30.0  38500.0   Diesel  90.0       0.0          0  2000      3   
...     ...   ...      ...      ...   ...       ...        ...   ...    ...   
1431   7500  69.0  20544.0   Petrol  86.0       1.0          0  1300      3   
1432  10845  72.0  19000.0   Petrol  86.0       0.0          0  1300      3   
1433   8500  71.0  17016.0   Petrol  86.0       0.0          0  1300      3   
1434   7250  70.0  16916.0   Petrol  86.0       1.0          0  1300      3   
1435   6950  76.0      NaN   Petrol   NaN       0.0          0  1600      5   

      Weight  
0       1165  
1       1165  
2     

#### CHECKING THE MISSING DATA

In [183]:
print(cars_data3.isnull().sum())

Price         0
Age           8
KM            5
FuelType      5
HP           13
MetColor      5
Automatic     0
CC            0
Doors         0
Weight        0
dtype: int64


#### Dataframe = Dataframe.apply(lambda x: x.fillna(x.mean()) if x.dtype == "float" or x.dtype == 'int' else x.fillna(x.value_counts().index[0])) 

In [184]:
cars_data3 = cars_data3.apply(lambda x: 
                              x.fillna(x.mean()) if (x.dtype == "float" or x.dtype == 'int')
                              else x.fillna(x.value_counts().index[0]))

In [185]:
print(cars_data3)
print(cars_data3.isnull().sum())

      Price        Age            KM FuelType         HP  MetColor  Automatic  \
0      6500  23.000000  46986.000000   Petrol   90.00000  1.000000          0   
1     13750  23.000000  68672.035639   Diesel   90.00000  0.674354          0   
2     13950  24.000000  41711.000000   Diesel  101.50246  1.000000          0   
3     14950  56.027311  48000.000000   Diesel   90.00000  0.000000          0   
4     13750  30.000000  38500.000000   Diesel   90.00000  0.000000          0   
...     ...        ...           ...      ...        ...       ...        ...   
1431   7500  69.000000  20544.000000   Petrol   86.00000  1.000000          0   
1432  10845  72.000000  19000.000000   Petrol   86.00000  0.000000          0   
1433   8500  71.000000  17016.000000   Petrol   86.00000  0.000000          0   
1434   7250  70.000000  16916.000000   Petrol   86.00000  1.000000          0   
1435   6950  76.000000  68672.035639   Petrol  101.50246  0.000000          0   

        CC  Doors  Weight  